File: src/python/command_registry.py
"""CommandRegistry - mirrors CommandRegistry.cpp."""

# ── CommandId enum ────────────────────────────────────────────────────────────
class CommandId:
    Unknown = "unknown"
    Start = "start"
    Pause = "pause"
    Stop = "stop"
    Resume = "resume"
    Reset = "reset"
    Target = "target"
    MaxRpm = "max_rpm"
    Freerun = "freerun"
    Direction = "direction"
    StopNextHigh = "stop_next_high"
    StopNextLow = "stop_next_low"
    EndPos = "end_pos"
    EndPosTurns = "end_pos_turns"
    Rodage = "rodage"
    RodageStop = "rodage_stop"
    RodageDist = "rodage_dist"
    RodagePasses = "rodage_passes"
    GeomStartTrim = "geom_start_trim"
    GeomEndTrim = "geom_end_trim"
    GeomStartTrimNudge = "geom_start_trim_nudge"
    GeomEndTrimNudge = "geom_end_trim_nudge"
    GeomPreset = "geom_preset"
    GeomTotal = "geom_total"
    GeomBottom = "geom_bottom"
    GeomTop = "geom_top"
    GeomMargin = "geom_margin"
    GeomWire = "geom_wire"
    GeomTppOffset = "geom_tpp_ofs"
    GeomScatter = "geom_scatter"
    WindowShift = "window_shift"
    WindingStyle = "winding_style"
    WindingSeed = "winding_seed"
    WindingLayerJitter = "winding_layer_jitter"
    WindingLayerSpeed = "winding_layer_speed"
    WindingHumanTraverse = "winding_human_traverse"
    WindingHumanSpeed = "winding_human_speed"
    WindingFirstPassTraverse = "winding_first_pass_traverse"
    RecipeImport = "recipe_import"
    LatOffset = "lat_offset"

    # All IDs for reverse lookup
    ALL = None  # filled below


# ── Value kinds ───────────────────────────────────────────────────────────────
KIND_NONE = "none"
KIND_INT = "integer"
KIND_FLOAT = "float"
KIND_BOOL = "boolean"
KIND_ENUM = "enum"
KIND_JSON = "json"

_COMMANDS = [
    (CommandId.Start, KIND_NONE, None, None, None, None, "", True, "Start winding session."),
    (CommandId.Pause, KIND_NONE, None, None, None, None, "", True, "Pause winding session."),
    (CommandId.Stop, KIND_NONE, None, None, None, None, "", True, "Stop winding and return to idle."),
    (CommandId.Resume, KIND_NONE, None, None, None, None, "", True, "Resume from paused state."),
    (CommandId.Reset, KIND_NONE, None, None, None, None, "", True, "Reset turns and return to idle."),
    (CommandId.Target, KIND_INT, 1, 200000, None, None, "", True, "Target turns count."),
    (CommandId.MaxRpm, KIND_INT, 10, 1500, None, None, "", True, "Max spindle RPM."),
    (CommandId.Freerun, KIND_BOOL, None, None, None, None, "true,false", False, "Freerun mode."),
    (CommandId.Direction, KIND_ENUM, None, None, None, None, "cw,ccw", False, "Spindle direction."),
    (CommandId.StopNextHigh, KIND_NONE, None, None, None, None, "", True, "Stop lateral at next high bound."),
    (CommandId.StopNextLow, KIND_NONE, None, None, None, None, "", True, "Stop lateral at next low bound."),
    (CommandId.EndPos, KIND_ENUM, None, None, None, None, "none,high,low", False, "Final carriage position."),
    (CommandId.EndPosTurns, KIND_INT, 1, 20, None, None, "", False, "Hold turns on final position."),
    (CommandId.Rodage, KIND_NONE, None, None, None, None, "", True, "Start rodage mode."),
    (CommandId.RodageStop, KIND_NONE, None, None, None, None, "", True, "Stop rodage mode."),
    (CommandId.RodageDist, KIND_FLOAT, None, None, 5.0, 100.0, "", False, "Rodage shuttle distance (mm)."),
    (CommandId.RodagePasses, KIND_INT, 1, 200, None, None, "", False, "Rodage pass count."),
    (CommandId.GeomStartTrim, KIND_FLOAT, None, None, -5.0, 5.0, "", False, "Low bound trim (mm)."),
    (CommandId.GeomEndTrim, KIND_FLOAT, None, None, -5.0, 5.0, "", False, "High bound trim (mm)."),
    (CommandId.GeomStartTrimNudge, KIND_FLOAT, None, None, -1.0, 1.0, "", False, "Nudge low bound trim."),
    (CommandId.GeomEndTrimNudge, KIND_FLOAT, None, None, -1.0, 1.0, "", False, "Nudge high bound trim."),
    (CommandId.GeomPreset, KIND_INT, 0, 20, None, None, "", False, "Apply geometry preset."),
    (CommandId.GeomTotal, KIND_FLOAT, None, None, 0.0, 200.0, "", False, "Total bobbin width (mm)."),
    (CommandId.GeomBottom, KIND_FLOAT, None, None, 0.0, 50.0, "", False, "Bottom flange width (mm)."),
    (CommandId.GeomTop, KIND_FLOAT, None, None, 0.0, 50.0, "", False, "Top flange width (mm)."),
    (CommandId.GeomMargin, KIND_FLOAT, None, None, 0.0, 20.0, "", False, "Winding margin (mm)."),
    (CommandId.GeomWire, KIND_FLOAT, None, None, 0.01, 1.0, "", False, "Wire diameter (mm)."),
    (CommandId.GeomTppOffset, KIND_INT, -2000, 2000, None, None, "", False, "Turns-per-pass offset."),
    (CommandId.GeomScatter, KIND_FLOAT, None, None, 0.5, 5.0, "", False, "Scatter factor."),
    (CommandId.WindowShift, KIND_FLOAT, None, None, -5.0, 5.0, "", True, "Shift both winding bounds (mm)."),
    (CommandId.WindingStyle, KIND_ENUM, None, None, None, None, "straight,scatter,human", False, "Pattern style."),
    (CommandId.WindingSeed, KIND_INT, 1, 2147483647, None, None, "", False, "Pattern random seed."),
    (CommandId.WindingLayerJitter, KIND_FLOAT, None, None, 0.0, 0.45, "", False, "Per-layer TPP jitter."),
    (CommandId.WindingLayerSpeed, KIND_FLOAT, None, None, 0.0, 0.45, "", False, "Per-layer speed jitter."),
    (CommandId.WindingHumanTraverse, KIND_FLOAT, None, None, 0.0, 0.45, "", False, "Smooth traverse variation."),
    (CommandId.WindingHumanSpeed, KIND_FLOAT, None, None, 0.0, 0.45, "", False, "Smooth speed variation."),
    (CommandId.WindingFirstPassTraverse, KIND_FLOAT, None, None, 0.40, 1.80, "", False, "First pass traverse scale."),
    (CommandId.RecipeImport, KIND_JSON, None, None, None, None, "", False, "Import full recipe JSON."),
    (CommandId.LatOffset, KIND_FLOAT, None, None, 0.0, 200.0, "", False, "Lateral home offset (mm)."),
]

# Build key-indexed dict
_CMD_BY_KEY: dict[str, dict] = {}
for (cid, kind, imin, imax, fmin, fmax, csv, mutable, help_) in _COMMANDS:
    _CMD_BY_KEY[cid] = {
        "id": cid,
        "key": cid,
        "kind": kind,
        "min_int": imin,
        "max_int": imax,
        "min_float": fmin,
        "max_float": fmax,
        "enum_csv": csv,
        "mutable_during_session": mutable,
        "help": help_,
    }

# Key aliases (matches normalizeKeyImpl in C++)
_KEY_ALIASES = {
    "max-rpm": "max_rpm",
    "windows_shift": "window_shift",
}


def _normalize_key(key: str) -> str:
    k = _KEY_ALIASES.get(key, key)
    # window_shift prefix normalization
    if k.startswith("window_shift"):
        k = "window_shift"
    return k


def _validate_value(defn: dict, value: str) -> bool:
    kind = defn["kind"]
    if kind == KIND_NONE:
        return value == ""
    if kind == KIND_BOOL:
        return value in ("true", "false")
    if kind == KIND_ENUM:
        return value in defn["enum_csv"].split(",")
    if kind == KIND_JSON:
        return len(value) >= 2 and value[0] == "{"
    if kind == KIND_INT:
        try:
            v = int(value)
            return defn["min_int"] <= v <= defn["max_int"]
        except (ValueError, TypeError):
            return False
    if kind == KIND_FLOAT:
        try:
            v = float(value)
            return defn["min_float"] <= v <= defn["max_float"]
        except (ValueError, TypeError):
            return False
    return False


class CommandRegistry:
    @staticmethod
    def find_by_key(key: str) -> dict | None:
        k = _normalize_key(key)
        return _CMD_BY_KEY.get(k)

    @staticmethod
    def normalize_and_validate(cmd: str, value: str) -> tuple[str, str] | None:
        """Returns (canonical_key, command_id) or None if invalid."""
        k = _normalize_key(cmd)
        defn = _CMD_BY_KEY.get(k)
        if defn is None:
            return None
        if not _validate_value(defn, value):
            return None
        return (k, defn["id"])

    @staticmethod
    def all() -> list[dict]:
        return list(_CMD_BY_KEY.values())

    @staticmethod
    def capabilities_json() -> str:
        import json
        from config import PICKUP_WS_PROTOCOL_VERSION, PICKUP_RECIPE_FORMAT_VERSION
        cmds = [
            {
                "key": d["key"],
                "type": d["kind"],
                "mutableDuringSession": d["mutable_during_session"],
                "help": d["help"],
            }
            for d in _CMD_BY_KEY.values()
        ]
        payload = {
            "capabilitiesVersion": "1",
            "protocol": {
                "ws": PICKUP_WS_PROTOCOL_VERSION,
                "recipe": PICKUP_RECIPE_FORMAT_VERSION,
            },
            "commands": cmds,
        }
        return json.dumps(payload, separators=(",", ":"))

File: resources/esp32/src/CommandRegistry.cpp
#include "CommandRegistry.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "Protocol.h"

namespace {

static const CommandDefinition kCommands[] = {
    { CommandId::Start, "start", CommandValueKind::None, 0, 0, 0, 0, "", true, "Start winding session." },
    { CommandId::Pause, "pause", CommandValueKind::None, 0, 0, 0, 0, "", true, "Pause winding session." },
    { CommandId::Stop, "stop", CommandValueKind::None, 0, 0, 0, 0, "", true, "Stop winding and return to idle." },
    { CommandId::Resume, "resume", CommandValueKind::None, 0, 0, 0, 0, "", true, "Resume from paused state." },
    { CommandId::Reset, "reset", CommandValueKind::None, 0, 0, 0, 0, "", true, "Reset turns and return to idle." },
    { CommandId::Target, "target", CommandValueKind::Integer, 1, 200000, 0, 0, "", true, "Target turns count." },
    { CommandId::MaxRpm, "max_rpm", CommandValueKind::Integer, 10, 1500, 0, 0, "", true, "Max spindle RPM." },
    { CommandId::Freerun, "freerun", CommandValueKind::Boolean, 0, 0, 0, 0, "true,false", false, "Freerun mode." },
    { CommandId::Direction, "direction", CommandValueKind::Enum, 0, 0, 0, 0, "cw,ccw", false, "Spindle direction." },
    { CommandId::StopNextHigh, "stop_next_high", CommandValueKind::None, 0, 0, 0, 0, "", true, "Stop lateral at next high bound." },
    { CommandId::StopNextLow, "stop_next_low", CommandValueKind::None, 0, 0, 0, 0, "", true, "Stop lateral at next low bound." },
    { CommandId::EndPos, "end_pos", CommandValueKind::Enum, 0, 0, 0, 0, "none,high,low", false, "Final carriage position." },
    { CommandId::EndPosTurns, "end_pos_turns", CommandValueKind::Integer, 1, 20, 0, 0, "", false, "Hold turns on final position." },
    { CommandId::Rodage, "rodage", CommandValueKind::None, 0, 0, 0, 0, "", true, "Start rodage mode." },
    { CommandId::RodageStop, "rodage_stop", CommandValueKind::None, 0, 0, 0, 0, "", true, "Stop rodage mode." },
    { CommandId::RodageDist, "rodage_dist", CommandValueKind::Float, 0, 0, 5.0f, 100.0f, "", false, "Rodage shuttle distance (mm)." },
    { CommandId::RodagePasses, "rodage_passes", CommandValueKind::Integer, 1, 200, 0, 0, "", false, "Rodage pass count." },
    { CommandId::GeomStartTrim, "geom_start_trim", CommandValueKind::Float, 0, 0, -5.0f, 5.0f, "", false, "Low bound trim (mm)." },
    { CommandId::GeomEndTrim, "geom_end_trim", CommandValueKind::Float, 0, 0, -5.0f, 5.0f, "", false, "High bound trim (mm)." },
    { CommandId::GeomStartTrimNudge, "geom_start_trim_nudge", CommandValueKind::Float, 0, 0, -1.0f, 1.0f, "", false, "Nudge low bound trim." },
    { CommandId::GeomEndTrimNudge, "geom_end_trim_nudge", CommandValueKind::Float, 0, 0, -1.0f, 1.0f, "", false, "Nudge high bound trim." },
    { CommandId::GeomPreset, "geom_preset", CommandValueKind::Integer, 0, 20, 0, 0, "", false, "Apply geometry preset." },
    { CommandId::GeomTotal, "geom_total", CommandValueKind::Float, 0, 0, 0.0f, 200.0f, "", false, "Total bobbin width (mm)." },
    { CommandId::GeomBottom, "geom_bottom", CommandValueKind::Float, 0, 0, 0.0f, 50.0f, "", false, "Bottom flange width (mm)." },
    { CommandId::GeomTop, "geom_top", CommandValueKind::Float, 0, 0, 0.0f, 50.0f, "", false, "Top flange width (mm)." },
    { CommandId::GeomMargin, "geom_margin", CommandValueKind::Float, 0, 0, 0.0f, 20.0f, "", false, "Winding margin (mm)." },
    { CommandId::GeomWire, "geom_wire", CommandValueKind::Float, 0, 0, 0.01f, 1.0f, "", false, "Wire diameter (mm)." },
    { CommandId::GeomTppOffset, "geom_tpp_ofs", CommandValueKind::Integer, -2000, 2000, 0, 0, "", false, "Turns-per-pass offset." },
    { CommandId::GeomScatter, "geom_scatter", CommandValueKind::Float, 0, 0, 0.5f, 5.0f, "", false, "Scatter factor." },
    { CommandId::WindowShift, "window_shift", CommandValueKind::Float, 0, 0, -5.0f, 5.0f, "", true, "Shift both winding bounds (mm)." },
    { CommandId::WindingStyle, "winding_style", CommandValueKind::Enum, 0, 0, 0, 0, "straight,scatter,human", false, "Pattern style." },
    { CommandId::WindingSeed, "winding_seed", CommandValueKind::Integer, 1, 2147483647, 0, 0, "", false, "Pattern random seed." },
    { CommandId::WindingLayerJitter, "winding_layer_jitter", CommandValueKind::Float, 0, 0, 0.0f, 0.45f, "", false, "Per-layer TPP jitter." },
    { CommandId::WindingLayerSpeed, "winding_layer_speed", CommandValueKind::Float, 0, 0, 0.0f, 0.45f, "", false, "Per-layer speed jitter." },
    { CommandId::WindingHumanTraverse, "winding_human_traverse", CommandValueKind::Float, 0, 0, 0.0f, 0.45f, "", false, "Smooth traverse variation." },
    { CommandId::WindingHumanSpeed, "winding_human_speed", CommandValueKind::Float, 0, 0, 0.0f, 0.45f, "", false, "Smooth speed variation." },
    { CommandId::WindingFirstPassTraverse, "winding_first_pass_traverse", CommandValueKind::Float, 0, 0, 0.40f, 1.80f, "", false, "First pass traverse scale." },
    { CommandId::RecipeImport, "recipe_import", CommandValueKind::Json, 0, 0, 0, 0, "", false, "Import full recipe JSON." },
    { CommandId::LatOffset, "lat_offset", CommandValueKind::Float, 0, 0, 0.0f, 200.0f, "", false, "Lateral home offset (mm)." },
};

bool isBool(const char* s) {
    return strcmp(s, "true") == 0 || strcmp(s, "false") == 0;
}

bool csvContains(const char* csv, const char* value) {
    if (!csv || !value) return false;
    const size_t vlen = strlen(value);
    const char* p = csv;
    while (*p) {
        const char* c = strchr(p, ',');
        const size_t len = c ? (size_t)(c - p) : strlen(p);
        if (len == vlen && strncmp(p, value, len) == 0) return true;
        if (!c) break;
        p = c + 1;
    }
    return false;
}

const CommandDefinition* findWindowShiftDef() {
    for (size_t i = 0; i < (sizeof(kCommands) / sizeof(kCommands[0])); ++i) {
        if (strcmp(kCommands[i].key, "window_shift") == 0) return &kCommands[i];
    }
    return nullptr;
}

bool normalizeKeyImpl(const char* in, char* out, size_t outLen) {
    if (!in || !out || outLen == 0) return false;

    if (strcmp(in, "max-rpm") == 0) in = "max_rpm";
    if (strcmp(in, "windows_shift") == 0) in = "window_shift";

    if (strncmp(in, "window_shift", strlen("window_shift")) == 0) {
        in = "window_shift";
    }

    const size_t len = strlen(in);
    if (len + 1 > outLen) return false;
    memcpy(out, in, len + 1);
    return true;
}

bool validateValue(const CommandDefinition* def, const char* value) {
    if (!def || !value) return false;

    switch (def->kind) {
    case CommandValueKind::None:
        return value[0] == '\0';

    case CommandValueKind::Boolean:
        return isBool(value);

    case CommandValueKind::Enum:
        return csvContains(def->enumValuesCsv, value);

    case CommandValueKind::Integer: {
        char* end = nullptr;
        const long v = strtol(value, &end, 10);
        if (!end || *end != '\0') return false;
        return v >= def->minInt && v <= def->maxInt;
    }

    case CommandValueKind::Float: {
        char* end = nullptr;
        const float v = strtof(value, &end);
        if (!end || *end != '\0') return false;
        return v >= def->minFloat && v <= def->maxFloat;
    }

    case CommandValueKind::Json:
        return value[0] == '{' && strlen(value) >= 2;
    }

    return false;
}

const char* kindToString(CommandValueKind kind) {
    switch (kind) {
    case CommandValueKind::None: return "none";
    case CommandValueKind::Integer: return "integer";
    case CommandValueKind::Float: return "float";
    case CommandValueKind::Boolean: return "boolean";
    case CommandValueKind::Enum: return "enum";
    case CommandValueKind::Json: return "json";
    default: return "unknown";
    }
}

} // namespace

const CommandDefinition* CommandRegistry::all(size_t& count) {
    count = sizeof(kCommands) / sizeof(kCommands[0]);
    return kCommands;
}

const CommandDefinition* CommandRegistry::findByKey(const char* key) {
    if (!key) return nullptr;

    if (strncmp(key, "window_shift", strlen("window_shift")) == 0) {
        return findWindowShiftDef();
    }

    for (size_t i = 0; i < (sizeof(kCommands) / sizeof(kCommands[0])); ++i) {
        if (strcmp(kCommands[i].key, key) == 0) return &kCommands[i];
    }
    return nullptr;
}

const CommandDefinition* CommandRegistry::findById(CommandId id) {
    for (size_t i = 0; i < (sizeof(kCommands) / sizeof(kCommands[0])); ++i) {
        if (kCommands[i].id == id) return &kCommands[i];
    }
    return nullptr;
}

bool CommandRegistry::normalizeAndValidate(const char* cmdIn,
                                           const char* valueIn,
                                           char* canonicalCmdOut,
                                           size_t canonicalCmdOutLen,
                                           CommandId* idOut) {
    if (!cmdIn || !valueIn) return false;
    if (!normalizeKeyImpl(cmdIn, canonicalCmdOut, canonicalCmdOutLen)) return false;

    const CommandDefinition* def = findByKey(canonicalCmdOut);
    if (!def) return false;
    if (!validateValue(def, valueIn)) return false;

    if (idOut) *idOut = def->id;
    return true;
}

void CommandRegistry::capabilitiesJson(char* outBuf, size_t outBufLen) {
    if (!outBuf || outBufLen == 0) return;

    size_t used = 0;
    int n = snprintf(outBuf, outBufLen,
                     "{\"capabilitiesVersion\":\"%s\","
                     "\"protocol\":{\"ws\":\"%s\",\"uart\":\"%s\",\"recipe\":%u},"
                     "\"commands\":[",
                     PICKUP_CAPABILITIES_VERSION,
                     PICKUP_WS_PROTOCOL_VERSION,
                     PICKUP_UART_PROTOCOL_VERSION,
                     (unsigned)PICKUP_RECIPE_FORMAT_VERSION);
    if (n < 0) {
        outBuf[0] = '\0';
        return;
    }
    used = (size_t)n;

    size_t count = 0;
    const CommandDefinition* defs = all(count);
    for (size_t i = 0; i < count && used < outBufLen; ++i) {
        const CommandDefinition& d = defs[i];
        n = snprintf(outBuf + used, outBufLen - used,
                     "%s{\"key\":\"%s\",\"type\":\"%s\",\"mutableDuringSession\":%s,\"help\":\"%s\"}",
                     (i == 0) ? "" : ",",
                     d.key,
                     kindToString(d.kind),
                     d.mutableDuringSession ? "true" : "false",
                     d.help ? d.help : "");
        if (n < 0) break;
        used += (size_t)n;
    }

    if (used + 3 <= outBufLen) {
        snprintf(outBuf + used, outBufLen - used, "]}");
    } else {
        // Truncated: close the JSON array to keep the output parseable.
        if (outBufLen >= 3) {
            outBuf[outBufLen - 3] = ']';
            outBuf[outBufLen - 2] = '}';
        }
        outBuf[outBufLen - 1] = '\0';
    }
}
